### PR TITLE
Clear resources during long training

### PR DESCRIPTION
**Pull request by evg-tyurin**
* Coach.py: recreate NNet after each iteration to clear resources
(especially in Keras+TF environment)
* Coach.py: free mcts variables after they are not used
* NeuralNet.py: abstract method for recreating models
* Keras implementation of Othello: reference method implementations for
recreating models

### CODE REVIEW

File: Coach.py [hunk 1]
```diff
@@ -101,9 +107,15 @@ def learn(self):
...
-                self.nnet = pnet
+                # self.nnet = pnet
+                self.nnet.recreate(folder=self.args.checkpoint, filename='temp.pth.tar')
 
             else:
                 print('ACCEPTING NEW MODEL')
                 self.nnet.save_checkpoint(folder=self.args.checkpoint, filename='checkpoint_' + str(i) + '.pth.tar')
-                self.nnet.save_checkpoint(folder=self.args.checkpoint, filename='best.pth.tar')                
+                self.nnet.save_checkpoint(folder=self.args.checkpoint, filename='best.pth.tar')
+                self.nnet.recreate(folder=self.args.checkpoint, filename='best.pth.tar')
```
Comment (suragnair) [inline Coach.py]: I don't think we need to recreate self.nnet here

File: Coach.py [hunk 2]
```diff
@@ -84,6 +84,9 @@ def learn(self):
                                                                                                            total=bar.elapsed_td, eta=bar.eta_td)
                 bar.next()
             bar.finish()
+            
+            # to be garbage collected asap
+            self.mcts = None
 
             # training new network, keeping a copy of the old one
             self.nnet.save_checkpoint(folder=self.args.checkpoint, filename='temp.pth.tar')
```
Comment (suragnair) [inline Coach.py]: We might have to recreate pnet in the line below.

**Comment by suragnair** [comment]
I just realised there's an easier way to do this. We can just define a `self.pnet` in the `__init__` of Coach. All subsequent updates can be through `load_checkpoint`. I think the main issue is this line `pnet = self.nnet.__class__(self.game)` since it introduces a new graph and session each time, which is not required.

**Comment by suragnair** [comment]
f391999 should solve it. Have a look.

File: Coach.py (hunk 1 above)
Comment (1424667164) [reply Coach.py]: I guess just clear `trainExamples` queue is all we need.

**Pull request closed without merging**